akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -378,14 +408,19 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
       override def onAllDataFlushed(): Unit = completeIfDone()
 
       private def completeIfDone(): Unit = {
-        val noMoreOutgoingStreams = isClosed(substreamIn) && activeStreamCount() == 0
+        val noMoreOutgoingStreams = (terminating || isClosed(substreamIn)) && activeStreamCount() == 0
         def allOutgoingDataFlushed = isClosed(frameOut) || multiplexer.hasFlushedAllData
         if (noMoreOutgoingStreams && allOutgoingDataFlushed) {
-          cancel(frameIn)
-          complete(frameOut)
-          // Using complete here (instead of a simpler `completeStage`) will make sure the buffer can be
-          // drained by the user before completely shutting down the stage finally.
-          bufferedSubStreamOutput.complete()
+          if (isServer)
+            completeStage()
```

Comment:
so the server never has pending things in `bufferedSubStreamOutput` here? That's not obvious to me.

Reply:
You are right, that's probably a bit optimistic.

Reply:
I checked again. On the server side, it would mean that a user handler would still have a *request* buffered. However, in that case `activeStreamCount` would be > 0. 

This is different on the client side, where all network traffic is done but the user might not have read final responses.

Reply:
Aah gotcha :+1: - might warrant a comment?

Reply:
Yep, adding one.